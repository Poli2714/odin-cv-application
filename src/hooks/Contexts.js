import { createContext } from 'react';

const IndexContext = createContext({
  activeIndex: null,
  setActiveIndex: () => {},
});

const KeyContext = createContext({
  activeKey: {
    eduActiveKey: null,
    expActiveKey: null,
  },
  setActiveKey: () => {},
});

const EduErrMsgContext = createContext({
  showEduErrMessage: false,
  setShowEduErrMessage: () => {},
});

const ExpErrMsgContext = createContext({
  showExpErrMessage: false,
  setShowExpErrMessage: () => {},
});

const PersonalInfoContext = createContext({
  personalInfo: {
    profilePicture: '',
    fullName: '',
    currentJobTitle: '',
    email: '',
    mobile: '',
    city: '',
  },
  setPersonalInfo: () => {},
});

const SummaryContext = createContext({
  summary: '',
  setSummary: () => {},
});

const ExperienceContext = createContext({
  experiences: {
    id: {
      jobTitle: '',
      employer: '',
      jobStartDate: '',
      jobEndDate: '',
      jobDescription: '',
    },
  },
  setExperiences: () => {},
});

const EducationContext = createContext({
  educations: {
    id: {
      degree: '',
      university: '',
      eduStartDate: '',
      eduEndDate: '',
      eduInfo: '',
    },
  },
  setEducations: () => {},
});

export {
  EducationContext,
  EduErrMsgContext,
  ExperienceContext,
  ExpErrMsgContext,
  IndexContext,
  KeyContext,
  PersonalInfoContext,
  SummaryContext,
};
