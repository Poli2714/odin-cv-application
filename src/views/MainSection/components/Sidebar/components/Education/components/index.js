import Degree from './Degree/Degree';
import EduEndDate from './EduEndDate/EduEndDate';
import EduInfo from './EduInfo/EduInfo';
import EduStartDate from './EduStartDate/EduStartDate';
import University from './University/University';

export { Degree, EduEndDate, EduInfo, EduStartDate, University };
