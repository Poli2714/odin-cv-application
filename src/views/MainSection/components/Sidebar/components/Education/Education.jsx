import { useContext } from 'react';

import {
  Degree,
  EduEndDate,
  EduInfo,
  EduStartDate,
  University,
} from './components';
import { DateInputsWrapper, FormWrapper } from 'src/components/layout';
import { Form } from 'src/components/forms';
import { Group } from 'src/components/ui';
import { EducationContext, KeyContext } from 'src/hooks/Contexts';

function Education() {
  const { educations } = useContext(EducationContext);
  const { activeKey, setActiveKey } = useContext(KeyContext);
  const { eduActiveKey } = activeKey;

  const handleClickEduGroup = (key, currentKey) =>
    currentKey !== key
      ? setActiveKey({ ...activeKey, eduActiveKey: key })
      : setActiveKey({
          ...activeKey,
          eduActiveKey: null,
        });

  const renderForm = key => (
    <Form key={key} isActive={eduActiveKey === key}>
      <Degree id={key} />
      <University id={key} />
      <DateInputsWrapper>
        <EduStartDate id={key} />
        <EduEndDate id={key} />
      </DateInputsWrapper>
      <EduInfo id={key} />
    </Form>
  );

  return (
    <>
      {Object.keys(educations).length === 1
        ? renderForm(Object.keys(educations)[0])
        : Object.keys(educations).map((key, i) => (
            <FormWrapper key={key}>
              <Group
                isActive={key !== eduActiveKey}
                label={`Education #${i + 1}`}
                onClick={() => handleClickEduGroup(key, eduActiveKey)}
              />
              {eduActiveKey === key && renderForm(key)}
            </FormWrapper>
          ))}
    </>
  );
}

export default Education;
