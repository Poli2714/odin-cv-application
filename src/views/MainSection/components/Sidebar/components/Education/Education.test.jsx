import { expect, test, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useState } from 'react';

import Education from './Education';
import { EducationContext, KeyContext } from 'src/hooks/Contexts';

vi.mock('src/components/layout', () => ({
  FormWrapper: vi.fn(({ children }) => (
    <div data-testid="form-wrapper">{children}</div>
  )),
  DateInputsWrapper: vi.fn(() => null),
}));

vi.mock('src/components/forms', () => ({
  Form: vi.fn(() => (
    <form action="" data-testid="form" method="get">
      <label htmlFor="test">Mock label</label>
      <input type="text" id="test" />
    </form>
  )),
}));

vi.mock('src/components/ui', () => ({
  Group: vi.fn(({ isActive, label, onClick }) => (
    <div
      className={isActive ? 'inactive' : 'active'}
      data-testid="group"
      onClick={onClick}
    >
      <span data-testid="group-label">{label}</span>
    </div>
  )),
}));

const data1 = { fakeID: '' };
const data2 = { fakeID1: '', fakeID2: '', fakeID3: '' };

function MockParentComponent({ eduData, id }) {
  const [educations] = useState(eduData);
  const [activeKey, setActiveKey] = useState({ eduActiveKey: id });

  return (
    <EducationContext.Provider value={{ educations }}>
      <KeyContext.Provider value={{ activeKey, setActiveKey }}>
        <Education />
      </KeyContext.Provider>
    </EducationContext.Provider>
  );
}

test('renders Education with single eduData', () => {
  render(<MockParentComponent eduData={data1} id="fakeID" />);

  expect(screen.queryByTestId('form-wrapper')).not.toBeInTheDocument();
  expect(screen.getByTestId('form')).toMatchSnapshot();
  expect(screen.getAllByTestId('form')).toHaveLength(1);
});

test('renders Education with multiple eduData where only second form is active', () => {
  render(<MockParentComponent eduData={data2} id="fakeID2" />);
  const groups = screen.getAllByTestId('group');
  const groupLabels = screen.getAllByTestId('group-label');

  expect(groups).toHaveLength(3);
  groupLabels.forEach((label, i) => {
    expect(label).toHaveTextContent(`Education #${i + 1}`);
  });
  groups.forEach((group, i) => {
    i === 1
      ? expect(group).toHaveClass('active')
      : expect(group).toHaveClass('inactive');
  });
});

test('displays only three Groups without any form when a user clicks on an active Group', async () => {
  const user = userEvent.setup();

  render(<MockParentComponent eduData={data2} id="fakeID3" />);
  const groups = screen.getAllByTestId('group');
  const activeGroup = groups[2];

  expect(groups).toHaveLength(3);
  expect(activeGroup).toHaveClass('active');
  expect(screen.getByTestId('form')).toBeInTheDocument();

  await user.click(activeGroup);
  groups.forEach(group => {
    expect(group).toHaveClass('inactive');
  });
  expect(screen.queryByTestId('form')).not.toBeInTheDocument();
});
