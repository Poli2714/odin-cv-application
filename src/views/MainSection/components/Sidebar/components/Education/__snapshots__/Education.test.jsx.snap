// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders Education with single eduData 1`] = `
<form
  action=""
  data-testid="form"
  method="get"
>
  <label
    for="test"
  >
    Mock label
  </label>
  <input
    id="test"
    type="text"
  />
</form>
`;
